src/network/networkd-dhcp6.c

```diff
@@ -35,6 +35,105 @@ static bool dhcp6_get_prefix_delegation(Link *link) {
+                                        subnet_id,
+                                        n_prefixes);
+                        return -1;
+                } else {
+                        r = in_addr_prefix_nth(AF_INET6, &prefix, 64, subnet_id);
+                        if (r < 0)
+                                return r;
+
+                        (void) in_addr_to_string(AF_INET6, &prefix, &assigned_buf);
+
+                        // Verify that the prefix we did calculate fits in the
+                        // pd prefix.  FIXME: this error should never happen
+                        // given that the code above is correct.
+                        r = in_addr_prefix_covers(AF_INET6, &pd_prefix_union, pd_prefix_len, &prefix);
+                        if (!r) {
+
+                                log_link_error(link, "Prefix %s is outside of the PD prefix. Not assigning.",
+                                                strnull(assigned_buf));
+                                return -1;
+                        }
+
+
+                        Link* assigned_link = dhcp6_prefix_get(manager, &prefix.in6);
+
+                        if (assigned_link && assigned_link != link) {
+                                log_link_error(link, "The requested prefix %s is already assigned to another link: %s",
+                                               assigned_buf,
+                                               strnull(assigned_link->ifname)
+                                );
+                                return -EAGAIN; // FIXME: is -EAGAIN the right error code here?
+                        }
+                        *addr = prefix.in6;
+
+                        log_link_debug(link, "The requested prefix %s is available. Using it.",
+                                       strnull(assigned_buf)
+                        );
+                        return 0;
+                }
+        } else {
+                for (uint64_t n  = 0; n < n_prefixes; n++) {
+                        // if we do not have an allocation preference just iterate
+                        // through the address space and return the first free prefix.
+                        Link* assigned_link = dhcp6_prefix_get(manager, &prefix.in6);
+
+                        if (!assigned_link || assigned_link == link) {
+                                *addr = prefix.in6;
+                                return 0;
+                        }
+
+                        r = in_addr_prefix_next(AF_INET6, &prefix, 64);
+                        // Running out of prefixes?
+                        if (r < 0) {
+                                log_link_error(link, "Can't allocate another prefix. Out of address space?");
+                                return r;
+                        }
+                }
+
+                log_link_warning(link, "Couldn't find a suitable prefix. Ran out of address space.");
+        }
+
+        return -1;
```

Comment:
please return a proper error code. basically, "return -1" is never acceptable in our code to indicate errors. Always needs to be "return -EAFNOSUPPORT" or "return -EINVAL" or whatnot.